examples/src/demo4.cxx

```diff
@@ -0,0 +1,128 @@
+//
+//
+/*!
+  \file
+  \ingroup examples
+  \brief An example of a method to compute the objective function value
+  of an image. All parameters are parsed from a parameter file.
+
+  It illustrates
+	- basic class derivation principles
+	- how to use ParsingObject to have automatic capabilities of parsing
+	  parameters files (and interactive questions to the user)
+	- how to initialise and setup a objective function object
+
+  Note that the same functionality could be provided without deriving
+  a new class from ParsingObject. One could have a KeyParser object
+  in main() and fill it in directly.
+
+  See README.txt in the directory where this file is located.
+
+  \author Kris Thielemans and Robert Twyman      
+*/
+/*
+    Copyright (C) 2004- 2012, Hammersmith Imanet Ltd
+
+    This software is distributed under the terms 
+    of the GNU General  Public Licence (GPL)
+    See STIR/LICENSE.txt for details
+*/
+
+#include "stir/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndProjData.h"
+#include "stir/IO/read_from_file.h"
+#include "stir/is_null_ptr.h"
+
+namespace stir {
+
+class MyStuff: public ParsingObject
+{
+public:
+  void set_defaults();
+  void initialise_keymap();
+  bool post_processing();
+  void run();
+  typedef DiscretisedDensity<3,float> target_type;
+
+protected:
+  shared_ptr<GeneralisedObjectiveFunction<target_type> >  objective_function_sptr;
+
+private:
+  std::string input_filename;
+  std::string image_filename;
+};
+
+void
+MyStuff::set_defaults()
+{
+  objective_function_sptr.reset(new PoissonLogLikelihoodWithLinearModelForMeanAndProjData<target_type>);
+}
+
+void 
+MyStuff::initialise_keymap()
+{
+  parser.add_start_key("MyStuff parameters");
+  parser.add_key("input file", &input_filename);
+  parser.add_key("image filename", &image_filename);
+  parser.add_parsing_key("objective function type", &objective_function_sptr);
+  parser.add_stop_key("End");
+}
+
+bool MyStuff::
+post_processing()
+{
+  if (is_null_ptr(this->objective_function_sptr))
+  {
+      error("objective_function_sptr is null");
+      return true;
+  }
+  return false;
+}
+
+void
+MyStuff::run()
+{
+
+  std::cout << "input_filename: " << input_filename << "\n";
+  std::cout << "image_filename: " << image_filename << "\n";
+
+  /////// load data from file
+  shared_ptr<DiscretisedDensity<3,float> > 
+    density_sptr(read_from_file<DiscretisedDensity<3,float> >(image_filename));
+
+  // Checks density_sptr is loaded correctly.
+  if (is_null_ptr(density_sptr))
+  {
+      error("density_sptr is null");
+  }
```

Comment:
Is this required? Will the `if (is_null_ptr(density_sptr))  = False `

Reply:
Do you mean "will `read_from_file` throw an exception or return a 0 pointer? I can't recall. Check the documentation... (or the code...), or just try!

if it is required, best to give a more informative message.